Dockerfile and docs to run containerized playbooks

codificat opened the pull request:
A Dockerfile based on [playbook2image](https://github.com/aweiteka/playbook2image) to containerize openshift-ansible so we can run an arbitrary playbook from a container.

WIP, want to gather feedback and thoughts around it.

DanyC97 commented:
@codificat for my own knowledge (sorry if is an obvious answer) if i understand correctly you want to run the `openshift-ansible` playbooks from containers running inside Openshift? if so i am a bit confused what is the benefit.

detiber commented on images/run-playbook/README.md:
```diff
@@ -0,0 +1,43 @@
+
+## Build
+
+1. **Docker**:
```
Is this capable of being run as system container? @giuseppe

giuseppe replied:
the configuration for the container seems easy enough that should not be a problem to make a system container of it we will need to provide the configuration for runC.

We'll probably benefit of running the systemd service as Type=oneshot

aweiteka replied:
This sounds really interesting. @detiber could you describe the use case a bit?

detiber replied:
@aweiteka The main use case is being able to run the installer on one of the hosts that is part of the installation.

The issue we have today is that as part of the installation we are making changes to the docker config which require a service restart. If the installer is running in a container under docker, then it goes *poof* when the docker restart happens.

aweiteka replied:
@codificat I'll look at putting the system container files into the playbook2image base so they're inherited for this optional use case. https://github.com/aweiteka/playbook2image/issues/22

codificat commented:
Some context: the trigger use case behind this was to be able to regularly run a playbook to check expiration dates of internal certificates using the openshift_certificate_expiry role. The idea was indeed to have this run from within OpenShift itself e.g. via a ScheduledJob. There are more use cases that would benefit from the "run a specific playbook from a container" approach.

There's another Dockerfile in the top level dir which has a different focus (i.e. to config/upgrade the cluster). One point to review is if it's worth unifying both types of use cases under a single image.

rhcarvalho commented:
@DanyC97 another use case and trigger for an earlier work on this that didn't make it into master was to run maintenance and health check playbooks in environments that do not have Ansible 2.2. In some corporate envs there are barries to upgrade/install RPMs, so the Docker image would allow us to ship a "snapshot" of openshift-ansible in a know-to-work configuration -- and that could be run as a Docker container, independent of OpenShift (some playbooks are useful before OpenShift is even installed).

detiber commented:
> There's another Dockerfile in the top level dir which has a different focus (i.e. to config/upgrade the cluster). One point to review is if it's worth unifying both types of use cases under a single image.

I would say the answer to this is: absolutely.

DanyC97 commented:
thank you @codificat @rhcarvalho for providing insights, much appreciated !

detiber commented:
@codificat looks like this PR is raising some yamllint errors: https://travis-ci.org/openshift/openshift-ansible/jobs/202362727#L953

detiber commented:
travis failures should clean up with a rebase, just fixed in master a few min ago.

detiber commented on Dockerfile:
```diff
@@ -1,26 +1,21 @@
-LABEL Architecture="x86_64"
-LABEL io.k8s.description="Ansible code and playbooks for installing Openshift Container Platform." \
-      io.k8s.display-name="Openshift Installer" \
-      io.openshift.tags="openshift,installer"
```
Might be good to have relevant metadata for this image, maybe @tdawson can help with recommendations.

codificat commented:
@detiber thanks for the feedback. yamllint should be happier now and I'll be rebasing, squashing etc as needed as this progresses.

About unifying into a single Dockerfile:
1. I'd move the existing `Dockerfile` to `Dockerfile.rhel7` as I believe it's only a prototype so far and no one is really using it yet, and it seems to be designed for productization (please correct me if that's not the case)
2. Moving the proposed `Dockerfile` that uses `playbook2image` to replace the current Dockerfile. Also moving the doc to the top level dir and not adding an `images` dir.

I'm still thinking how to best handle input/output for the general playbook case. The current Dockerfile uses/requires a VOLUME, not sure if that's the one-and-only way...

jcantrill commented:
@DanyC97 'oc cluster up' also has need to deploy logging and metrics in order that we may fully deprecate their deployers

sosiouxme commented on Dockerfile:
```diff
@@ -1,26 +1,21 @@
+# Setting a default of openshift_facts.yml as it's an informative playbook
+# that can help test that everything is set properly (inventory, sshkeys)
+ENV PLAYBOOK_FILE=plabyooks/byo/openshift_facts.yml \
+    ANSIBLE_HOST_KEY_CHECKING=False \
```
I get why this is needed for non-interactive usage, but could there be a comment as to why?

Something to the effect that unless you mount in a pre-verified known_hosts file, there's no security mechanism to verify the host keys at the time this is running. While we could use the method in http://stackoverflow.com/a/35213839/262768 to set up such a file at run-time, it wouldn't offer any actual security.

detiber replied:
Can we provide a user to run the system container as and automatically mount ~/.ssh for that user?

Then we could provide a flag to opt in for host key checking.

This would also help for ssh proxy config, environment-specific auth, etc.

codificat replied:
Right... for the use cases of running the built image from inside OpenShift itself I was thinking to use a ConfigMap for these settings. For a standalone container I'll update the docs to explain that better.

One thing to keep in mind with mounting `~/.ssh` is that perms/selinux come into play. That's why the example `docker run` in `README_CONTAINER_IMAGE` sets the uid with `id -u` and mounts the ssh key with `:Z` - this has a side effect or leaving your key labelled something like `system_u:object_r:container_file_t:s0:c609,c894`

sosiouxme commented on README_CONTAINER_IMAGE.md:
```diff
@@ -0,0 +1,37 @@
+1. Using standalone **Docker**:
+
+        cd openshift-ansible
+        docker build -t certcheck -f images/certificate-check/Dockerfile .
```
errr... the images/certificate-check/Dockerfile doesn't seem to exist...?

codificat replied:
Rigth, I moved it to the top level dir and forgot to update the doc, will fix, thanks for spotting!

sosiouxme commented on README_CONTAINER_IMAGE.md:
```diff
@@ -0,0 +1,37 @@
+           -e PLAYBOOK_FILE=playbooks/certificate_expiry/default.yaml \
+           certcheck
+
+There [playbook2image examples](https://github.com/aweiteka/playbook2image/tree/master/examples) provide additional information on how to use a built image.
```
s/There/The/

codificat commented:
Addressed the comments above: added some labels, fixed typos/path references and removed `ANSIBLE_HOST_KEY_CHECKING`.

It turns out that the `ansible.cfg` in the repo contains  `host_key_checking=False`. It's worth documenting this more and explain how to add `known_hosts` to the container. Maybe we should remove that setting from `ansible.cfg`? Not sure why it's there as a default

Will update `Dockerfile.rhel7` after some tests with OSBS.

codificat commented:
Rebased and squashed.

@detiber PTAL

There are things to be improved (more docs, including ssh key handling; rhel7 Dockerfile update; input/output handling) but these can be done in another PR... having this one in should give us a base to build upon, including having images built and tested.

rhcarvalho commented on Dockerfile.rhel7:
```diff
@@ -2,16 +2,17 @@ FROM rhel7
       io.k8s.display-name="Openshift Installer" \
       io.openshift.tags="openshift,installer"
 
+
```
nit: extra line?

codificat replied:
ditto, will defer changes to the product Dockerfile

rhcarvalho commented on Dockerfile.rhel7:
```diff
@@ -2,16 +2,17 @@ FROM rhel7
+LABEL version="v3.4.1.8"
+LABEL release="1"
+LABEL com.redhat.component="aos3-installation-docker"
+LABEL architecture="x86_64"
```
Good to update these, though we need to make sure this format will already be used by the build tools?
I remember once when we updated labels and then had to revert because the build tool was not accepting the new standard at the time.

codificat replied:
Right - some changes will be necessary to this Dockerfile but there are other steps to do first, most notably to have a base playbook2image in the build system.
I'll revert that commit and leave Dockerfile.rhel7 untouched (just a rename of the previous Dockerfile)

rhcarvalho commented on Dockerfile:
```diff
@@ -1,26 +1,31 @@
+# including documentation for the settings/env vars referenced below
+FROM docker.io/aweiteka/playbook2image:latest
 
-MAINTAINER Troy Dawson <tdawson@redhat.com>
```
Shall we declare a MAINTAINER (perhaps a group/team) for this new image?

codificat replied:
Will add one and fix the typos :)

sosiouxme replied:
nope, you're on the hook forever!

I suggest dev@lists.openshift.redhat.com for all our maintainers.

rhcarvalho replied:
@sosiouxme that's the course of action we took, see the line below :)

rhcarvalho replied:
The original file was moved to `Dockerfile.rhel7`, but because of the name this file is getting compared to the old one.

codificat replied:
Right, I applied the changes and force-pushed, but somehow github was still showing these outdated comments... the latest commit (`c6b43ef`) has dev@lists.openshift.redhat.com as maintainer.

rhcarvalho commented on Dockerfile:
```diff
@@ -1,26 +1,31 @@
-MAINTAINER Troy Dawson <tdawson@redhat.com>
+LABEL name="openshift-ansible" \
+      summary="OpenShift's installation and configuration tool" \
+      description="A containerized openshift-ansible image to let you run playbooks to install, upgrade, maintain, chek, etc. an OpenShift cluster" \
```
nit: typo s/chek/check

I'd probably remove the "etc", for it doesn't add information for users/customers.

rhcarvalho commented on Dockerfile:
```diff
@@ -1,26 +1,31 @@
+      description="A containerized openshift-ansible image to let you run playbooks to install, upgrade, maintain, chek, etc. an OpenShift cluster" \
+      url="https://github.com/openshift/openshift-ansible" \
+      io.k8s.display-name="openshift-ansible" \
+      io.k8s.description="A containerized openshift-ansible image to let you run playbooks to install, upgrade, maintain, chek, etc. an OpenShift cluster" \
```
Same as above

rhcarvalho commented on README_CONTAINER_IMAGE.md:
```diff
@@ -0,0 +1,38 @@
+At the very least, when running a container using an image built this way you must specify:
+
+1. The **playbook** to run. This is set using the `PLAYBOOK_FILE` environment variable.
+1. An **inventory** file. One way to provide this is to serve the inventory file from a web server and use the `INVENTORY_URL` environment variable to fetch it.
```
I'd suggest to document the use of INVENTORY_FILE instead.

Users will already need to learn/deal with mounts for SSH keys, so I think using that for the inventory would be more natural than have people figure out how to serve their file over HTTP.

codificat replied:
Right, will do that - I had problems when I tried this before and that's why I resorted to url-based, leaving that for later, but it works now, so I'll update the docs.

rhcarvalho commented:
aos-ci-test
